CactbotOverlay/FFXIVProcess.cs

```diff
@@ -789,17 +916,42 @@ public class NinjaJobData {
       }
     }
 
-    public NinjaJobData GetNinja() {
-      byte[] bytes = GetJobSpecificData();
-      if (bytes == null)
-        return null;
-
-      var j = new NinjaJobData();
-      j.huton_ms = BitConverter.ToUInt32(bytes, kJobDataInnerStructOffsetJobSpecificData);
-      j.ninki_amount = bytes[kJobDataInnerStructOffsetJobSpecificData + 5];
-      return j;
+    public unsafe NinjaJobData GetNinja() {
+      fixed (byte* p = GetJobSpecificData()) {
+        if (p == null)
+          return null;
+        else {
+          NinjaJobMemory mem = *(NinjaJobMemory*)&p[0];
+          NinjaJobData j = new NinjaJobData() {
+            huton_ms = mem.huton_ms,
+            ninki_amount = mem.ninki_amount,
+          };
+          return j;
+        }
+      }
     }
 
+    [StructLayout(LayoutKind.Explicit)]
+    public struct BlackMageJobMemory {
+      [FieldOffset(0x00)]
+      public ushort polyglot_time_ms; // Number of ms left before polygot proc.
+
+      [FieldOffset(0x02)]
+      public ushort umbral_time_ms; // Number of ms left in umbral fire/ice.
+
+      [FieldOffset(0x04)]
+      public ushort umbral_state; // Positive = Umbral Fire Stacks, Negative = Umbral Ice Stacks.
+
+      [FieldOffset(0x05)]
+      public byte umbral_hearts_count;
+
+      [FieldOffset(0x06)]
+      public byte foul_count;
+
+      [FieldOffset(0x07)]
+      public byte enochian_state; // Bit 0 = Enochian active. Bit 1 = Polygot active.
```

Comment:
A few of the classes job data needs further unpacking before it can be passed off which makes the JobData classes still necessary as far as I am aware.

Reply:
I'm not totally sure I agree here, but I think that can be for a followup pull request and not done here.  This is all a really nice change as it is.  For what it's worth, I think these memory classes could just have a ToJObject() function that can do the unpacking as it needs.